src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.QueueItem.cs

```diff
@@ -5,59 +5,94 @@
+            /// Begins executing the work specified by this queue item.
+            /// Note that this does not take in a cancellation token as the queue item
+            /// itself already has a combined cancellation token that is used.
             /// </summary>
-            private readonly Func<RequestContext?, CancellationToken, Task> _callbackAsync;
+            public Task CallbackAsync(RequestContext? context);
 
             /// <summary>
-            /// <see cref="CorrelationManager.ActivityId"/> used to properly correlate this work with the loghub
-            /// tracing/logging subsystem.
+            /// Sets the result of this queue item to an exception.  Used if we fail to enqueue this item before it starts.
             /// </summary>
-            public readonly Guid ActivityId;
-            private readonly ILspLogger _logger;
+            public bool TrySetException(Exception e);
 
-            /// <inheritdoc cref="IRequestHandler.MutatesSolutionState" />
-            public readonly bool MutatesSolutionState;
+            public CancellationToken CombinedCancellationToken { get; }
 
             /// <inheritdoc cref="IRequestHandler.RequiresLSPSolution" />
-            public readonly bool RequiresLSPSolution;
+            public bool RequiresLSPSolution { get; }
+
+            /// <inheritdoc cref="IRequestHandler.MutatesSolutionState" />
+            public bool MutatesSolutionState { get; }
 
             /// <inheritdoc cref="RequestContext.ClientName" />
-            public readonly string? ClientName;
-            public readonly string MethodName;
+            public string? ClientName { get; }
 
-            /// <inheritdoc cref="RequestContext.ClientCapabilities" />
-            public readonly ClientCapabilities ClientCapabilities;
+            public string MethodName { get; }
 
             /// <summary>
             /// The document identifier that will be used to find the solution and document for this request. This comes from the <see cref="TextDocumentIdentifier"/> returned from the handler itself via a call to <see cref="IRequestHandler{RequestType, ResponseType}.GetTextDocumentIdentifier(RequestType)"/>.
             /// </summary>
-            public readonly TextDocumentIdentifier? TextDocument;
+            public TextDocumentIdentifier? TextDocument { get; }
+
+            /// <inheritdoc cref="RequestContext.ClientCapabilities" />
+            public ClientCapabilities ClientCapabilities { get; }
+
+            /// <summary>
+            /// <see cref="CorrelationManager.ActivityId"/> used to properly correlate this work with the loghub
+            /// tracing/logging subsystem.
+            /// </summary>
+            public Guid ActivityId { get; }
+
+            public RequestMetrics Metrics { get; }
+        }
+
+        private readonly struct QueueItem<TRequestType, TResponseType> : IQueueItem
```

Comment:
being a struct likely isn't helpful (since it is presumably boxed into an IQueueItem.  i tend to find using a struct in that case more confusing as i now have to think about boxing.  consdier making a class (or convince me that a struct is the right thing here).